here-naksha-cli/src/jvmMain/java/com/here/naksha/cli/copy/service/CopyService.java

```
@@ -48,36 +53,114 @@ private List<NakshaFeature> readFeaturesFromSrc(
-            case SuccessResponse successResponse -> extractResponseItems(successResponse, NakshaFeature.class);
-            case ErrorResponse errorResponse -> throw new CopyServiceException(
-                    "Problem with reading from source!",
-                    new NakshaException(errorResponse.getError())
-            );
-            default -> throw new CopyServiceException("Unexpected response from source!");
-        };
+        SuccessResponse successResponse = requireSourceSuccessResponse(response);
+        return extractResponseItems(successResponse, NakshaFeature.class);
     }
 
-    private SuccessResponse writeFeaturesToTarget(
+    private List<String> writeFeaturesToTarget(
             List<NakshaFeature> features,
-            CopyElement target
+            CopyElement target,
+            boolean autoCreateTarget
     ) throws CopyServiceException {
+        List<String> messages = new ArrayList<>();
         IStorage storage = useTargetStorage(target);
-        WriteRequest writeRequest = createFeaturesRequest(
-                target.getMapId(),
-                target.getCollectionId(),
-                features
+        if (autoCreateTarget) {
+            List<String> createTargetMessages = createTarget(storage, target);
+            messages.addAll(createTargetMessages);
+        }
+        Response response = performCreateFeaturesRequest(storage, target, features);
+        requireTargetSuccessResponse(response);
+        return messages;
+    }
+
+    private Response performCreateFeaturesRequest(
+            IStorage storage, CopyElement target, List<NakshaFeature> features
+    ) throws CopyServiceException {
+        WriteRequest addFeaturesRequest = createFeaturesRequest(target.getMapId(), target.getCollectionId(), features);
+        return performWriteRequest(storage, addFeaturesRequest);
+    }
+
+    private List<String> createTarget(IStorage storage, CopyElement target) throws CopyServiceException {
```

We should validate and fail early if `mapId` / `collectionId` are missing

---

ok